src/tree.rs

```diff
@@ -548,39 +627,79 @@ where
+            (Some(child_node_key), mut siblings) => {
+                let (update_result, proof_opt) = self.insert_at(
+                    child_node_key,
+                    version,
+                    nibble_iter,
+                    value,
+                    tree_cache,
+                    with_proof,
+                )?;
+
+                let new_proof_opt = proof_opt.map(|proof| {
+                    // The move siblings function allows zero copy moves for proof
+                    let proof_leaf = proof.leaf();
+                    let mut new_siblings = proof.move_siblings();
+                    new_siblings.append(&mut siblings);
+                    SparseMerkleProof::new(proof_leaf, new_siblings)
+                });
+
+                (update_result, new_proof_opt)
             }
-            None => {
+            (None, siblings) => {
                 if let Some(value) = value {
                     // insert
                     let new_child_node_key = node_key.gen_child_node_key(version, child_index);
-                    PutResult::Updated(Self::create_leaf_node(
-                        new_child_node_key,
-                        nibble_iter,
-                        value,
-                        tree_cache,
-                    )?)
+
+                    if with_proof {
+                        // The Merkle proof doesn't have a leaf
+                        let merkle_proof = SparseMerkleProof::new(None, siblings);
+                        (
+                            PutResult::Updated(Self::create_leaf_node(
+                                new_child_node_key,
+                                nibble_iter,
+                                value,
+                                tree_cache,
+                            )?),
+                            Some(merkle_proof),
+                        )
+                    } else {
+                        (
+                            PutResult::Updated(Self::create_leaf_node(
+                                new_child_node_key,
+                                nibble_iter,
+                                value,
+                                tree_cache,
+                            )?),
+                            None,
+                        )
+                    }
                 } else {
-                    // delete not found
-                    PutResult::NotChanged
+                    if with_proof {
+                        let merkle_proof = SparseMerkleProof::new(None, siblings);
+                        // delete not found
+                        (PutResult::NotChanged, Some(merkle_proof))
+                    } else {
+                        (PutResult::NotChanged, None)
+                    }
                 }
             }
         };
 
         // Reuse the current `InternalNode` in memory to create a new internal node.
         let mut children: Children = internal_node.into();
-        match result {
+        match result.0 {
```

Comment:
nit: I think this would be more readable if we destructured the tuple. 

```rust
let (put_result, proof) = result;
match put_result { 
// ...
```